ulib/axlibc/src/socket.rs

```diff
@@ -1,542 +1,199 @@
 #[no_mangle]
-pub unsafe extern "C" fn ax_accept(
+pub unsafe extern "C" fn accept(
     socket_fd: c_int,
     socket_addr: *mut ctypes::sockaddr,
     socket_len: *mut ctypes::socklen_t,
-) -> ctypes::ssize_t {
-    debug!(
-        "ax_accept <= {} {:#x} {:#x}",
-        socket_fd, socket_addr as usize, socket_len as usize
-    );
-    ax_call_body!(ax_accept, {
-        if socket_addr.is_null() || socket_len.is_null() {
-            return Err(LinuxError::EFAULT);
-        }
-        let socket = Socket::from_fd(socket_fd)?;
-        let new_socket = socket.accept()?;
-        let addr = new_socket.peer_addr()?;
-        let new_fd = Socket::add_to_fd_table(Socket::Tcp(Mutex::new(new_socket)))?;
-        unsafe {
-            (*socket_addr, *socket_len) = into_sockaddr(addr);
-        }
-        Ok(new_fd)
-    })
+) -> c_int {
+    e(sys_accept(socket_fd, socket_addr, socket_len))
 }
 
 /// Shut down a full-duplex connection.
 ///
 /// Return 0 if success.
 #[no_mangle]
-pub unsafe extern "C" fn ax_shutdown(
+pub unsafe extern "C" fn shutdown(
     socket_fd: c_int,
     flag: c_int, // currently not used
-) -> ctypes::ssize_t {
-    debug!("ax_shutdown <= {} {}", socket_fd, flag);
-    ax_call_body!(ax_shutdown, {
-        Socket::from_fd(socket_fd)?.shutdown()?;
-        Ok(0)
-    })
+) -> c_int {
+    e(sys_shutdown(socket_fd, flag))
 }
 
 /// Query addresses for a domain name.
 ///
+/// Only IPv4. Ports are always 0. Ignore service and hint.
+/// Results' ai_flags and ai_canonname are 0 or NULL.
+///
 /// Return address number if success.
 #[no_mangle]
-pub unsafe extern "C" fn ax_getaddrinfo(
+pub unsafe extern "C" fn getaddrinfo(
     node: *const c_char,
     service: *const c_char,
-    addrs: *mut ctypes::sockaddr,
-    len: ctypes::size_t,
+    _hints: *const ctypes::addrinfo,
+    res: *mut *mut ctypes::addrinfo,
 ) -> c_int {
-    let name = char_ptr_to_str(node);
-    let port = char_ptr_to_str(service);
-    debug!(
-        "ax_getaddrinfo <= {:?} {:?} {:#x} {}",
-        name, port, addrs as usize, len
-    );
-    ax_call_body!(ax_getaddrinfo, {
-        if addrs.is_null() || (node.is_null() && service.is_null()) {
-            return Err(LinuxError::EFAULT);
-        }
-        let addr_slice = unsafe { core::slice::from_raw_parts_mut(addrs, len) };
-        let res = if let Ok(domain) = name {
-            if let Ok(a) = domain.parse::<IpAddr>() {
-                vec![a]
-            } else {
-                axnet::dns_query(domain)?
-            }
-        } else {
-            vec![Ipv4Addr::LOCALHOST.into()]
-        };
-
-        for (i, item) in res.iter().enumerate().take(len) {
-            addr_slice[i] = into_sockaddr(SocketAddr::new(
-                *item,
-                port.map_or(0, |p| p.parse::<u16>().unwrap_or(0)),
-            ))
-            .0;
-        }
-        Ok(res.len().min(len))
-    })
+    let addrs = [ctypes::sockaddr::default(); ctypes::MAXADDRS as usize].as_mut_ptr();
```

Comment:
It must be allocated on the heap because that address will be used after the function returns.

Reply:
`Vec` will still be freed after the function returns.

Furthermore, the buffer allocated by `getaddrinfo` should be deallocated in `freeaddrinfo`, so `sys_freeaddrinfo` also needs to be implemented in Rust.

You can use `Box::into_raw()`/`Box::from_raw()` to allocate/deallocate the buffer in Rust.